lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1143,37 +1143,122 @@ bool SinkReshapes::run(Function *F, const CompilationContext &cctx) {
+        DCHECK(node->getNthResult(0).getType()->isFusedQuantizedType() ||
+               node->getNthInput(1).getType()->isFusedQuantizedType() ||
+               node->getNthResult(0).dims().equals(node->getNthInput(1).dims()))
+            << "SinkReshapes: not an element-wise node: " << node->toString();
+
+        // At least one of the inputs must be a Reshape.
+        // If both inputs are Reshapes, they must have the same dimensions.
+        auto *LRN = dyn_cast<ReshapeNode>(node->getNthInput(0));
+        auto *RRN = dyn_cast<ReshapeNode>(node->getNthInput(1));
+        if (!LRN && !RRN) {
+          continue;
+        }
+        if (LRN && RRN &&
+            !LRN->getResult().dims().equals(RRN->getResult().dims())) {
+          continue;
+        }
 
-    // Create new eltwise node.
-    auto in = RS->getInput();
-    auto out = node->getNthResult(0);
-    auto newTy =
-        F->getParent()->uniqueTypeWithNewShape(out.getType(), in.dims());
-    auto *newN = F->addNode(node->clone());
-    newN->setNthInput(0, in);
-    newN->setTypeUnsafe(0, newTy);
-    newN->setPredicate(node->getPredicate());
+        // Canonicalize node to simplify transformation implementation (make LHS
+        // always be the input with a Reshape).
+        bool swap = (LRN == nullptr);
+        auto nv = node->getNthInput(1);
+        if (swap) {
+          nv = node->getNthInput(0);
+          LRN = RRN;
+          RRN = nullptr;
+        }
 
-    // Create new Reshape.
-    auto *newRS = F->createReshape(RS->getName(), newN,
-                                   RS->getResult().getType()->dims());
-    newRS->setPredicate(node->getPredicate());
-    out.replaceAllUsesOfWith(newRS->getResult());
+        // RHS must be either a Reshape or a Constant (+ Quantize) or Splat.
+        auto *RQ = dyn_cast<QuantizeNode>(nv);
+        auto *RC = dyn_cast<Constant>(RQ ? RQ->getInput() : nv);
+        auto *RS = dyn_cast<SplatNode>(nv);
+        if (!RRN && !RC && !RS) {
+          continue;
+        }
 
-    changed = true;
+        // Create new Constant, Quantize or Splat, if needed.
+        NodeValue rhs;
+        if (RRN) {
+          rhs = RRN->getInput();
+        }
+        if (RC) {
+          auto ty = F->getParent()->uniqueTypeWithNewShape(
+              RC->getType(), LRN->getInput().dims());
+          auto *newC = F->getParent()->createConstant(ty, RC->getName());
+          newC->getPayloadMutable().copyRawFrom(&RC->getPayload());
+          rhs = newC->getOutput();
+        }
+        if (RQ) {
+          auto ty = F->getParent()->uniqueTypeWithNewShape(
+              RQ->getResult().getType(), LRN->getInput().dims());
+          rhs = F->createQuantize(RQ->getName(), rhs, ty);
+        }
+        if (RS) {
+          auto ty = F->getParent()->uniqueTypeWithNewShape(
+              RS->getResult().getType(), LRN->getInput().dims());
+          rhs = F->createSplat(RS->getName(), ty, RS->getValue());
+        }
```

Comment:
All this code seems unnecessarily complex. We have optimizations that do folding of Reshapes into Constants/Splats, and if we have a `Constant -> Quantize -> Reshape` then it'll get constant folded too. Can we not just add a Reshape here and let other optimizations clean it up later? E.g.

```suggestion
        // Reshape RHS to match LHS.
        NodeValue rhs = F->createSplat(nv.getNode()->getName(), nv, LRN->getInput().dims());
```